File: app/src/main/java/com/qimian233/ztool/hook/modules/safecenter/EnableAutorunByDefault.java
package com.qimian233.ztool.hook.modules.safecenter;

import com.qimian233.ztool.hook.base.BaseHookModule;

import io.github.libxposed.api.XposedInterface;
import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;

public class EnableAutorunByDefault extends BaseHookModule {
    public static final String FEATURE_NAME = "default_enable_autorun";

    private static final int ATTR_WHITELIST = 0x20000000;
    private static final int ATTR_RELATIVE_WHITELIST = 0x40000000;

    public EnableAutorunByDefault() {}

    @Override
    public String getModuleName() {
        return FEATURE_NAME;
    }

    @Override
    public String[] getTargetPackages() {
        return new String[] {"com.lenovo.safecenter", "com.zui.safecenter"};
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        String packageName = param.getPackageName();
        if ("com.zui.safecenter".equals(packageName) || "com.lenovo.safecenter".equals(packageName)) {
            logger.info("Start hooking safecenter");
            try {
                Class<?> cls = classLoader.loadClass("com.lenovo.performance.autorun.beans.AutoRunDbItem");
                Field fld = cls.getDeclaredField("mAttrs");
                fld.setAccessible(true);

                for (Constructor<?> ctor : cls.getDeclaredConstructors()) {
                    hookWithId(ctor, "ctor", chain -> {
                        chain.proceed();
                        Object obj = chain.getThisObject();
                        int attrs = fld.getInt(obj);
                        attrs |= ATTR_WHITELIST | ATTR_RELATIVE_WHITELIST;
                        fld.setInt(obj, attrs);
                        return null;
                    });
                }
                logger.info("Hooked safecenter [OK]");
            } catch (Exception e) {
                logger.error("Failed hooking safecenter", e);
            }
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/safecenter/EnableAutorunByDefault.kt
package com.qimian233.ztool.hook.modules.safecenter

import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.lang.reflect.Field

class EnableAutorunByDefault : AppHookModule() {

    override fun getModuleName(): String = PreferenceKeys.DEFAULT_ENABLE_AUTORUN.name

    override fun getTargetPackages(): Array<String> = arrayOf(
        ScopeKeys.LENOVO_SAFE_CENTER.packageName,
        ScopeKeys.ZUI_SAFE_CENTER.packageName
    )

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        val packageName = param.packageName
        if (ScopeKeys.ZUI_SAFE_CENTER.packageName == packageName
            || ScopeKeys.LENOVO_SAFE_CENTER.packageName == packageName
        ) {
            logger.info("Start hooking safecenter")
            try {
                val cls = classLoader.loadClass("com.lenovo.performance.autorun.beans.AutoRunDbItem")
                val fld: Field = cls.getDeclaredField("mAttrs")
                fld.isAccessible = true

                for (ctor in cls.declaredConstructors) {
                    hookWithId(ctor, "ctor") { chain ->
                        chain.proceed()
                        val obj = chain.thisObject
                        var attrs = fld.getInt(obj)
                        attrs = attrs or (ATTR_WHITELIST or ATTR_RELATIVE_WHITELIST)
                        fld.setInt(obj, attrs)
                        null
                    }
                }
                logger.info("Hooked safecenter [OK]")
            } catch (e: Exception) {
                logger.error("Failed hooking safecenter", e)
            }
        }
    }

    companion object {
        private const val ATTR_WHITELIST = 0x20000000
        private const val ATTR_RELATIVE_WHITELIST = 0x40000000
    }
}
